components/clp-package-utils/clp_package_utils/controller.py

```diff
@@ -0,0 +1,466 @@
+import json
+import logging
+import multiprocessing
+import os
+import pathlib
+import socket
+import stat
+import subprocess
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+from clp_py_utils.clp_config import (
+    AwsAuthType,
+    CLPConfig,
+    COMPRESSION_SCHEDULER_COMPONENT_NAME,
+    COMPRESSION_WORKER_COMPONENT_NAME,
+    DB_COMPONENT_NAME,
+    GARBAGE_COLLECTOR_COMPONENT_NAME,
+    QUERY_SCHEDULER_COMPONENT_NAME,
+    QUERY_WORKER_COMPONENT_NAME,
+    QUEUE_COMPONENT_NAME,
+    REDIS_COMPONENT_NAME,
+    REDUCER_COMPONENT_NAME,
+    RESULTS_CACHE_COMPONENT_NAME,
+    StorageEngine,
+    StorageType,
+    WEBUI_COMPONENT_NAME,
+)
+from clp_py_utils.clp_metadata_db_utils import (
+    get_archives_table_name,
+    get_datasets_table_name,
+    get_files_table_name,
+)
+
+from clp_package_utils.general import (
+    check_docker_dependencies,
+    CONTAINER_CLP_HOME,
+    generate_docker_compose_container_config,
+    get_clp_home,
+    validate_db_config,
+    validate_queue_config,
+    validate_redis_config,
+    validate_results_cache_config,
+    validate_webui_config,
+)
+
+LOGS_FILE_MODE = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IROTH
+
+logger = logging.getLogger(__name__)
+
+
+def _get_ip_from_hostname(hostname: str) -> str:
```

Comment:
need to convert hostnames to IP addresses because docker compose's `<service>.ports` mappings don't like hostnames